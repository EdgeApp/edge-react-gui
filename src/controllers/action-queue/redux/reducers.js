// @flow

import { type Reducer, combineReducers } from 'redux'

import { type Action } from '../../../types/reduxTypes'
import { type ActionQueueMap } from '../types'

export type ActionQueueState = {
  +queue: ActionQueueMap
}

export const actionQueue: Reducer<ActionQueueState, Action> = combineReducers({
  queue(state: ActionQueueMap = {}, action: Action): ActionQueueMap {
    switch (action.type) {
      case 'ACTION_QUEUE/LOAD_QUEUE': {
        return action.data
      }
      case 'ACTION_QUEUE/QUEUE_ITEM': {
        return {
          ...state,
          [action.programId]: action.item
        }
      }
      case 'ACTION_QUEUE/UPDATE_PROGRAM_STATE': {
        const { programId } = action.state
        const { effect } = action.state

        // Remove program from action queue if it has finished
        if (effect?.type === 'done') {
          const { [programId]: _removed, ...rest } = state
          return rest
        }

        return {
          ...state,
          [programId]: {
            ...state[programId],
            state: action.state
          }
        }
      }
      default:
        return state
    }
  }
})
