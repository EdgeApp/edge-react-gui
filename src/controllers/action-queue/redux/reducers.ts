import { combineReducers } from 'redux'

import type { Action } from '../../../types/reduxTypes'
import type { ActionQueueMap } from '../types'
import { checkEffectIsDone } from '../util/checkEffectIsDone'

export interface ActionQueueState {
  readonly actionQueueMap: ActionQueueMap
  readonly activeProgramIds: string[]
}

export const actionQueue = combineReducers<ActionQueueState, Action>({
  actionQueueMap(state: ActionQueueMap = {}, action: Action): ActionQueueMap {
    switch (action.type) {
      case 'ACTION_QUEUE/LOAD_QUEUE': {
        return action.data
      }
      case 'ACTION_QUEUE/QUEUE_ITEM': {
        return {
          ...state,
          [action.programId]: action.item
        }
      }
      case 'ACTION_QUEUE/UPDATE_PROGRAM_STATE': {
        const { programId } = action.state

        return {
          ...state,
          [programId]: {
            ...state[programId],
            state: action.state
          }
        }
      }
      case 'LOGOUT': {
        return {}
      }
      default:
        return state
    }
  },
  activeProgramIds(state: string[] = [], action: Action): string[] {
    switch (action.type) {
      case 'ACTION_QUEUE/LOAD_QUEUE': {
        const ids = Object.keys(action.data)
        return ids
      }
      case 'ACTION_QUEUE/QUEUE_ITEM': {
        return [...state, action.programId]
      }
      case 'ACTION_QUEUE/UPDATE_PROGRAM_STATE': {
        const { programId } = action.state
        const { effect } = action.state

        // Remove program from action queue if it has finished
        if (checkEffectIsDone(effect)) {
          return state.filter(id => id !== programId)
        }

        return state
      }
      case 'LOGOUT': {
        return []
      }
      default:
        return state
    }
  }
})
