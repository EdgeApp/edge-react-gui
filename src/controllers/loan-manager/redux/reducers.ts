import { combineReducers, Reducer } from 'redux'

import { Action } from '../../../types/reduxTypes'
import { LoanAccountMap } from '../types'

export type LoanManagerState = {
  readonly loanAccounts: LoanAccountMap
}

export const loanManager: Reducer<LoanManagerState, Action> = combineReducers({
  loanAccounts(state: LoanAccountMap = {}, action: Action): LoanAccountMap {
    switch (action.type) {
      case 'LOAN_MANAGER/SET_LOAN_ACCOUNT': {
        const { loanAccount } = action
        return {
          ...state,
          [loanAccount.id]: loanAccount
        }
      }
      case 'LOAN_MANAGER/DELETE_LOAN_ACCOUNT': {
        const { id } = action
        const { [id]: _, ...remainingState } = state
        return remainingState
      }
      case 'LOGOUT': {
        return {}
      }
      default:
        return state
    }
  }
})
