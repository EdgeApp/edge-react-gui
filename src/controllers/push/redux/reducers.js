// @flow

import { type Reducer, combineReducers } from 'redux'

import { type Action } from '../../../types/reduxTypes'
import { type PushEvent } from '../types'

export type PushState = {
  +pushEvents: PushEvent[]
}

export const actionQueue: Reducer<PushState, Action> = combineReducers({
  pushEvents(state: PushEvent[] = [], action: Action): PushEvent[] {
    switch (action.type) {
      case 'PUSH/ADD_PUSH_EVENTS': {
        return [...state, ...action.events]
      }
      default:
        return state
    }
  }
})
