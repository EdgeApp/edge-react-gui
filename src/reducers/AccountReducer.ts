import { combineReducers, Reducer } from 'redux'

import { Action } from '../types/reduxTypes'
import { AccountReferral, Promotion, ReferralCache } from '../types/ReferralTypes'

/**
 * App state tied to the core account object.
 */
export type AccountState = {
  readonly accountReferral: AccountReferral
  readonly accountReferralLoaded: boolean
  readonly referralCache: ReferralCache
  // TODO: Move account settings in here
}

const defaultAccountReferral: AccountReferral = {
  promotions: [],
  ignoreAccountSwap: false,
  hiddenAccountMessages: {}
}

const defaultReferralCache: ReferralCache = {
  accountMessages: [],
  accountPlugins: []
}

const accountInner: Reducer<AccountState, Action> = combineReducers({
  accountReferral(state: AccountReferral = defaultAccountReferral, action: Action): AccountReferral {
    switch (action.type) {
      case 'ACCOUNT_REFERRAL_LOADED': {
        const { referral } = action.data
        // Activating a promo link can race with the intial load,
        // so keep any promotions we already have:
        const promotions = mergePromotions(referral.promotions, state.promotions)
        return { ...referral, promotions }
      }
      case 'ACCOUNT_SWAP_IGNORED': {
        return { ...state, ignoreAccountSwap: action.data }
      }
      case 'MESSAGE_TWEAK_HIDDEN': {
        const { messageId, source } = action.data
        if (source.type === 'promotion') {
          const promotions = state.promotions.map(promo => {
            if (promo.installerId !== source.installerId) return promo
            const hiddenMessages = { ...promo.hiddenMessages, [messageId]: true }
            return { ...promo, hiddenMessages }
          })
          return { ...state, promotions }
        } else {
          const hiddenAccountMessages = { ...state.hiddenAccountMessages, [messageId]: true }
          return { ...state, hiddenAccountMessages }
        }
      }
      case 'PROMOTION_ADDED': {
        const newPromo = action.data
        const promotions = mergePromotions(state.promotions, [newPromo])
        return { ...state, promotions }
      }
      case 'PROMOTION_REMOVED': {
        const installerId = action.data
        const promotions = state.promotions.filter(promo => promo.installerId !== installerId)
        return { ...state, promotions }
      }
    }
    return state
  },

  accountReferralLoaded(state: boolean = false, action: Action): boolean {
    return action.type === 'ACCOUNT_REFERRAL_LOADED' ? true : state
  },

  referralCache(state: ReferralCache = defaultReferralCache, action: Action): ReferralCache {
    switch (action.type) {
      case 'ACCOUNT_REFERRAL_LOADED': {
        const { cache } = action.data
        return cache
      }
      case 'ACCOUNT_TWEAKS_REFRESHED': {
        const cache = action.data
        return cache
      }
    }
    return state
  }
})

// Shared logout logic:
export const account: Reducer<AccountState, Action> = (state: AccountState | undefined, action: Action) => {
  if (action.type === 'LOGOUT') {
    return accountInner(undefined, { type: 'DUMMY_ACTION_PLEASE_IGNORE' })
  }
  return accountInner(state, action)
}

/**
 * Merges lists of promotions, preferring items from b if there's a conflict.
 */
function mergePromotions(a: Promotion[], b: Promotion[]): Promotion[] {
  const toRemove: { [id: string]: true } = {}
  for (const promo of b) toRemove[promo.installerId] = true
  return [...a.filter(promo => !toRemove[promo.installerId]), ...b]
}
