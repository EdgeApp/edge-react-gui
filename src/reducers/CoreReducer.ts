import { Disklet } from 'disklet'
import { EdgeAccount, EdgeContext } from 'edge-core-js/types'
import { combineReducers } from 'redux'

import { Action } from '../types/reduxTypes'

export interface CoreState {
  readonly account: EdgeAccount
  readonly context: EdgeContext
  readonly disklet: Disklet
  readonly otpErrorShown: boolean
}

const flowHack: any = {}
const defaultContext: EdgeContext = flowHack
const defaultDisklet: Disklet = flowHack

const accountHack: any = {
  activeWalletIds: [],
  currencyConfig: {},
  currencyWallets: {},
  watch: () => () => {}
}
export const defaultAccount: EdgeAccount = accountHack

export const core = combineReducers<CoreState, Action>({
  account(state: EdgeAccount = defaultAccount, action: Action): EdgeAccount {
    switch (action.type) {
      case 'LOGIN':
        return action.data.account
      case 'LOGOUT':
        return defaultAccount
      default:
        return state
    }
  },

  context(state: EdgeContext = defaultContext, action: Action): EdgeContext {
    return action.type === 'CORE/CONTEXT/ADD_CONTEXT' ? action.data.context : state
  },

  disklet(state: Disklet = defaultDisklet, action: Action): Disklet {
    return action.type === 'CORE/CONTEXT/ADD_CONTEXT' ? action.data.disklet : state
  },

  otpErrorShown(state: boolean = false, action: Action): boolean {
    switch (action.type) {
      case 'OTP_ERROR_SHOWN':
        return true
      case 'LOGOUT':
        return false
      default:
        return state
    }
  }
})
