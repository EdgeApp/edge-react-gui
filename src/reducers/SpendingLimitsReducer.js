// @flow

import { type Reducer, combineReducers } from 'redux'

import { type Action } from '../types/reduxTypes.js'
import type { PluginSpendingLimits, SpendingLimits } from '../types/types'

export type SpendingLimitsState = {
  pluginLimits: { [key: string]: PluginSpendingLimits }
} & SpendingLimits

export const DEFAULT_PLUGIN_SPENDING_LIMITS: PluginSpendingLimits = {
  transaction: {
    isEnabled: false,
    amount: 0
  },
  fiatCurrencyCode: 'USD'
}

// REDUCERS
export const initialState = {
  transaction: {
    isEnabled: false,
    amount: 0
  },
  pluginLimits: {}
}

export const isEnabled = (state: boolean = initialState.transaction.isEnabled, action: Action) => {
  switch (action.type) {
    case 'ACCOUNT_INIT_COMPLETE':
    case 'SPENDING_LIMITS/NEW_SPENDING_LIMITS': {
      return action.data.spendingLimits.transaction.isEnabled
    }
    default:
      return state
  }
}

export const amount = (state: number = initialState.transaction.amount, action: Action) => {
  switch (action.type) {
    case 'ACCOUNT_INIT_COMPLETE':
    case 'SPENDING_LIMITS/NEW_SPENDING_LIMITS': {
      return action.data.spendingLimits.transaction.amount
    }
    default:
      return state
  }
}

export const transaction = combineReducers({
  isEnabled,
  amount
})

export const pluginLimits: Reducer<{ [key: string]: PluginSpendingLimits }, Action> = (state = initialState.pluginLimits, action: Action) => {
  switch (action.type) {
    case 'ACCOUNT_INIT_COMPLETE': {
      return {
        ...state,
        ...action.data.spendingLimits.pluginLimits
      }
    }
    case 'SPENDING_LIMITS/NEW_SPENDING_LIMITS': {
      if (!action.data.currencyCode || !action.data.fiatCurrencyCode) return state

      return {
        ...state,
        [action.data.currencyCode]: {
          ...action.data.spendingLimits,
          fiatCurrencyCode: action.data.fiatCurrencyCode
        }
      }
    }
    default:
      return state
  }
}

export const spendingLimits = combineReducers({
  transaction,
  pluginLimits
})
