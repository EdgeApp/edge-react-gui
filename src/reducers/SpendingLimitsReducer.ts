import { combineReducers } from 'redux'

import { Action } from '../types/reduxTypes'

// REDUCERS
export const initialState = {
  transaction: {
    isEnabled: false,
    amount: 0
  }
}

export const isEnabled = (state: boolean = initialState.transaction.isEnabled, action: Action) => {
  switch (action.type) {
    case 'ACCOUNT_INIT_COMPLETE':
    case 'SPENDING_LIMITS/NEW_SPENDING_LIMITS': {
      return action.data.spendingLimits.transaction.isEnabled
    }
    default:
      return state
  }
}

export const amount = (state: number = initialState.transaction.amount, action: Action) => {
  switch (action.type) {
    case 'ACCOUNT_INIT_COMPLETE':
    case 'SPENDING_LIMITS/NEW_SPENDING_LIMITS': {
      return action.data.spendingLimits.transaction.amount
    }
    default:
      return state
  }
}

export const transaction = combineReducers({
  isEnabled,
  amount
})

export const spendingLimits = combineReducers({
  transaction
})
