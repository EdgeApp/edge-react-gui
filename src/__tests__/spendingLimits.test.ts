import { describe, expect, it } from '@jest/globals'

import { initialState, spendingLimits } from '../reducers/SpendingLimitsReducer'

describe('spendingLimits', () => {
  it('should render initialState', () => {
    const actual = spendingLimits(undefined, {
      type: 'DUMMY_ACTION_PLEASE_IGNORE'
    })

    expect(actual).toMatchSnapshot()
  })

  describe('when logging in', () => {
    it('should update', () => {
      const actual = spendingLimits(initialState, {
        type: 'ACCOUNT_INIT_COMPLETE',
        data: {
          spendingLimits: {
            transaction: {
              isEnabled: false,
              amount: 150
            }
          }
        } as any
      })

      expect(actual).toMatchSnapshot()
    })
  })

  describe('from spending limits scene', () => {
    it('should disable', () => {
      const initialState = {
        transaction: {
          isEnabled: true,
          amount: 0
        }
      }
      const actual = spendingLimits(initialState, {
        type: 'SPENDING_LIMITS/NEW_SPENDING_LIMITS',
        data: {
          spendingLimits: {
            transaction: {
              isEnabled: false,
              amount: 234
            }
          }
        }
      })

      expect(actual).toMatchSnapshot()
    })

    it('should enable', () => {
      const initialState = {
        transaction: {
          isEnabled: false,
          amount: 0
        }
      }
      const actual = spendingLimits(initialState, {
        type: 'SPENDING_LIMITS/NEW_SPENDING_LIMITS',
        data: {
          spendingLimits: {
            transaction: {
              isEnabled: true,
              amount: 234
            }
          }
        }
      })

      expect(actual).toMatchSnapshot()
    })
  })
})
