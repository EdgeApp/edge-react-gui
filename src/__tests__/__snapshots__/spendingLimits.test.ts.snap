// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spendingLimits from spending limits scene should disable 1`] = `
Object {
  "transaction": Object {
    "amount": 234,
    "isEnabled": false,
  },
}
`;

exports[`spendingLimits from spending limits scene should enable 1`] = `
Object {
  "transaction": Object {
    "amount": 234,
    "isEnabled": true,
  },
}
`;

exports[`spendingLimits should render initialState 1`] = `
Object {
  "transaction": Object {
    "amount": 0,
    "isEnabled": false,
  },
}
`;

exports[`spendingLimits when logging in should update 1`] = `
Object {
  "transaction": Object {
    "amount": 150,
    "isEnabled": false,
  },
}
`;
